Build PyVelox Wheels for Linux and MacOS (x86_64)

assignUser opened the pull request:
This PR sets up a CI Job to build pyvelox wheels for x86_64 MacOS and Linux for python 3.9 - 3.11. The wheels are uploaded as artifacts. We can easily add a job to publish them to pypi regularly.

assignUser commented:
I don't really understand why the Mac build failed as it worked with the same commit on my fork https://github.com/assignUser/velox/actions/runs/4056432125 I'll re-run it...

assignUser commented:
It might make sense to reduce the scope of building to one version of python for PRs (or remove this from PRs in favour of #3924 ). @kgpai @pedroerp thoughts? And could you test the mac wheels?

assignUser commented:
(the failed docker build is releated to ubuntu repos timing out, see previous commits were it succeeded )

pedroerp commented on CMakeLists.txt:
```diff
@@ -318,7 +334,7 @@ find_package(gflags COMPONENTS shared)
 find_package(glog REQUIRED)
 
 set_source(fmt)
-resolve_dependency(fmt 8.0.1)
```
This seems a bit unrelated. What's this change about?

assignUser replied:
The torcharrow image had system fmt installed which was the old and incompatible version 6.x. This change sets a minimum version for fmt and prevents `resolve_dependency`  from picking up old system versions and instead build fmt from source, I used 8.0.1 as that is what folly uses (and the first version that has the internal v8 namespace)

assignUser commented:
@raulcd and me have tested the linux wheels successfully on ubuntu 22.04 and 20.04.  There seems to be a problem with the macos wheels due to the unusually high deployment target of 11 (required for preadv and pwritev) as some python binaries report to be a lower macos version due to build time settings vs what they actually are.
```python
from distutils import util
print(util.get_platform())
```
The above code will print the macos version if it says macos 10.xx you will not be able to install the wheel with the default name e.g. `pyvelox-0.0.1a0+bba0a2e-cp37-cp37m-macosx_11_0_x86_64.whl `  but you can rename the wheel to ...macos_10_15... and then you will be able to install it. Your system needs to support the BMI1/2 and AVX instruction sets of course, otherwise there will be a segfault as soon as you try to use the package (as is happening on the GHA macos runners)

assignUser commented:
On the GHA runners I was able to circumvent the issue by `brew install python@3.7`

assignUser commented:
I have added a publish step that only needs the right pypi API token added as a secret to work. An sdist will also be published.  Additionally the workflow can now be triggered manually for a custom ref with a custom version. 

If no version is given a def version will automatically generated based on the number of commits since the last release (or the first pyvelox commit). 

I also modified the workflow to work with ccache to speed up building of the wheels considerably.

assignUser commented:
I have added the README.md as the long description because twine was erroring otherwise https://github.com/assignUser/velox/actions/runs/4157573323/jobs/7192564200#step:6:16

I also updated the license field to Apache 2.0 as it was still set for BSD (I assume because it was copy pasted?)

kgpai commented on scripts/setup-macos.sh:
```diff
@@ -85,6 +85,12 @@ function install_fmt {
   cmake_install -DFMT_TEST=OFF
 }
 
+function install_folly {
```
shouldnt this be called somewhere ?

kgpai replied:
NVM : I can see now where its called..

kgpai commented on .github/workflows/build_pyvelox.yml:
```diff
@@ -0,0 +1,177 @@
+        uses: pypa/cibuildwheel@v2.12.0
+        env:
+          # required for preadv/pwritev
+          MACOSX_DEPLOYMENT_TARGET: "11.0"
```
nit: Is this why wheels would not install ? Can we set it to 10.9 to be more backwards compatible (?)

assignUser replied:
No due to the use of `preadv` and `pwritev`  11.0 is the lowest possible deployment target.

kgpai commented on setup.py:
```diff
@@ -158,10 +161,12 @@ def build_extension(self, ext):
     author="Meta",
     author_email="velox@fb.com",
-    license="BSD",
+    license="Apache License 2.0",
```
I think TA was released with BSD license - what is motivation for this change ?

assignUser replied:
Velox overall is licensed as Apache 2.0 :)

kgpai reviewed (commented):
Looks good ; some very minor questions/nits..

assignUser commented:
The logic for the version is not run on PRs so I created a test run here https://github.com/assignUser/test-repo-a/actions/runs/4199482457/jobs/7284415223
